### PR TITLE
Feature/setup authentication

### LINKED ISSUE #20: Setup user authentication
Allow user registration and login functionality.

### PR DESCRIPTION
**Pull request by vivekimsit**
- [x] Restructure server directory structure
- [x] Start server and client via same command `npm start`
- [x] Setup testing for server
- [x] Create sample `unit` and `e2e` tests
- [x] User authentication using `passport`

closes #20 #11 #8

### CODE REVIEW

File: models/user/user.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+const mongoose = require('mongoose');
+const { Schema } = mongoose;
+mongoose.Promise = global.Promise;
+const validator = require('validator');
+const mongodbErrorHandler = require('mongoose-mongodb-errors');
+const passportLocalMongoose = require('passport-local-mongoose');
+
+// Passport-Local Mongoose will add a username,
```
Comment (Mortuie) [inline models/user/user.js]: make it a multi line comment please `/* */`

File: web/app.js [hunk 2]
```diff
@@ -5,30 +5,34 @@ const express = require('express');
...
+app.use(bodyParser.urlencoded({ extended: true }));
+// Exposes a bunch of methods for validating data.
+// Used heavily on userController.validateRegister
+app.use(expressValidator());
+// Passport JS is what we use to handle our logins
+app.use(passport.initialize());
+app.use(passport.session());
 app.use(routes);
 // Serve static assets
 app.use(express.static(path.resolve(__dirname, '..', 'frontend', 'build')));
```
Comment (Mortuie) [inline web/app.js]: We need to talk about how we are going to handle this.

File: web/routes/index.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+const houses = require('./houses');
+const users = require('./users');
+const { catchAsyncErrors } = require('../utils');
+
+const router = new Router();
+
+router.get('/api/v1/houses', houses.list);
+router.post('/api/v1/houses', auth.isLoggedIn, catchAsyncErrors(houses.create));
+
+// 1. Validate the registration data
```
Comment (Mortuie) [inline web/routes/index.js]: multi line comment please.

File: web/app.js (hunk 2 above)
Comment (vivekimsit) [reply web/app.js]: 👍

File: web/routes/index.js (hunk 3 above)
Comment (vivekimsit) [reply web/routes/index.js]: Done

File: models/user/user.js (hunk 1 above)
Comment (vivekimsit) [reply models/user/user.js]: Done

**Comment by Mortuie** [comment]
I have a feeling I will get merge conflicts here with mine :-1:

**Pull request merged**